backend/backend/services/openAI/service.py

```diff
@@ -11,29 +21,82 @@ def generate_prompt(input):
+        import regex as re
+        graph_service = GraphService(protocol=subgraph)
+        schema = os.path.join(os.getcwdb().decode("utf-8"), graph_service.subgraph.schema_file_location)
+        mappings = os.path.join(os.getcwdb().decode("utf-8"), "/subgraphs/subgraphs/{}/src/".format(graph_service.subgraph.deployments['base']))
+        examples = os.getcwdb().decode("utf-8")+ "/backend/services/graph/graphql_examples.py"
+        #set recursive = True for case of uniswap etc where there are more sub directories
+        documents = SimpleDirectoryReader(input_files=[examples], recursive=True).load_data()
+        # print("documents", documents)
+        # save to disk
+        # index.save_to_disk('index.json')
+        # load from disk
+        # index = GPTSimpleVectorIndex.load_from_disk('index.json')
+
+        # define LLM
+        llm_predictor = LLMPredictor(llm=OpenAI(temperature=0, model_name="text-davinci-003"))
+
+        # define prompt helper
+        # set maximum input size
+        max_input_size = 4096
+        # set number of output tokens
+        num_output = 256
+        # set maximum chunk overlap
+        max_chunk_overlap = 20
+        prompt_helper = PromptHelper(max_input_size, num_output, max_chunk_overlap)
+
+        index = GPTSimpleVectorIndex(
+            documents, llm_predictor=llm_predictor, prompt_helper=prompt_helper
         )
-        openai_result = response.choices[0].text
+
+        # response = index.query(PRE_PROMPT.format(input_query))
+        prompt = GraphPromptFactory(subgraph).build_prompt_for_subgraph(input_query)
+        response = index.query(prompt)
+        # import pdb;pdb.set_trace()
+        openai_result = response.response
         print("==========openai response:==========\n", openai_result)
-        # strip any unnecessary text prepended and/or postpended to the gql query
-        openai_result = openai_result[
-            openai_result.find("{") : openai_result.rfind("}") + 1
-        ]
-        print("==========openai response (formatted):==========\n", openai_result)
         return openai_result
+
+    # def request_gql_for_graph(self, input_query, subgraph, generic=False):
+    #     prompt = input_query if generic else generate_prompt(input_query)
+    #     response = openai.Completion.create(
+    #         model="text-davinci-003",
+    #         prompt=generate_prompt(input_query),
+    #         temperature=0.6,
+    #         max_tokens=2048,
+    #     )
+    #     openai_result = response.choices[0].text
+    #     print("==========openai response:==========\n", openai_result)
+    #     # strip any unnecessary text prepended and/or postpended to the gql query
+    #     openai_result = openai_result[
+    #         openai_result.find("{") : openai_result.rfind("}") + 1
+    #     ]
+    #     print("==========openai response (formatted):==========\n", openai_result)
+    #     return openai_result
```

Comment:
would suggest to keep this in the implementation, sometimes gpt still gives weird unparsable graphql code back. stripping like this prevents the code from not being readable by the graph's api

Reply:
I was pondering about how to parse the gql, we could have tossed the schema into the client and validated a request on request.

Reply:
Maybe a formatter in the future